File: 20251220 - UI_MAP_BUILDER_INTEGRATION/topologies/hub_with_stepped_islands.py
# src/map_generator/topologies/hub_with_stepped_islands.py

import random
from .base_topology import BaseTopology
import copy
from src.map_generator.models.path_info import PathInfo, Coord
from src.utils.geometry import add_vectors, FORWARD_Z, FORWARD_X, BACKWARD_Z, BACKWARD_X, UP_Y, DOWN_Y, scale_vector
from typing import List, Dict, Any, Iterator

class HubWithSteppedIslandsTopology(BaseTopology):
    """
    Tạo ra một hòn đảo trung tâm và 4 hòn đảo vệ tinh ở các độ cao khác nhau,
    được nối với nhau bằng các bậc thang đi lên hoặc đi xuống.
    """

    def generate_path_info_variants(self, grid_size: tuple, params: dict, max_variants: int) -> Iterator[PathInfo]:
        """
        [KIẾN TRÚC MỚI] Tạo ra các biến thể với số lượng đảo và cấu trúc độ cao khác nhau.
        """
        count = 0
        
        # Các kịch bản về độ cao và số lượng đảo
        scenarios = [
            [3, -3],              # 2 đảo, 1 lên 1 xuống
            [2, -2, 4],           # 3 đảo, độ cao khác nhau
            [2, 3, 4],            # 3 đảo, tất cả đi lên
            [2, -2, 3, -3],       # 4 đảo
            [-2, -3, -4]          # 3 đảo, tất cả đi xuống
        ]

        for scenario in scenarios:
            if count >= max_variants:
                return
            
            # Tạo một bản sao sâu của params để không ảnh hưởng đến các lần lặp sau
            variant_params = copy.deepcopy(params)
            # Ghi đè height_options bằng kịch bản hiện tại
            variant_params['height_options'] = scenario
            
            yield self.generate_path_info(grid_size, variant_params)
            count += 1

    def _create_square_platform(self, center: Coord, size: int) -> list[Coord]:
        """Tạo một nền tảng hình vuông với tâm tại `center` và kích thước `size`."""
        cx, cy, cz = center
        half = size // 2
        platform_coords = []
        for x_offset in range(-half, half + 1):
            for z_offset in range(-half, half + 1):
                platform_coords.append((cx + x_offset, cy, cz + z_offset))
        return platform_coords

    def _create_staircase(self, start_point: Coord, direction: Coord, height_diff: int, gap_size: int) -> tuple[list[Coord], list[Coord]]:
        """
        [FIX] Tạo cầu thang với các bậc có thể nhảy được (1 height unit per step).
        `height_diff` dương thì đi lên, âm thì đi xuống.
        `direction` là hướng di chuyển trên mặt phẳng (X, Z).
        Returns:
            - path_coords: Đường đi trên cầu thang (tất cả các block đều có thể đi được).
            - obstacle_coords: Các khối vật cản (rỗng - không cần obstacles cho design này).
        """
        path = []
        obstacles = []
        current_pos = start_point
        
        # [FIX] Tạo cầu thang với các bậc có thể đi được
        # Mỗi bậc cao 1 unit (jumpable) và tiến 1 block theo direction
        steps_needed = abs(height_diff)
        y_step = 1 if height_diff > 0 else -1
        
        for _ in range(steps_needed):
            # Tiến theo hướng horizontal
            current_pos = add_vectors(current_pos, direction)
            # Thay đổi độ cao
            current_pos = (current_pos[0], current_pos[1] + y_step, current_pos[2])
            path.append(current_pos)
        
        # Thêm các bước gap ở độ cao cuối cùng (không thay đổi Y)
        for _ in range(gap_size - steps_needed):
            if gap_size > steps_needed:
                current_pos = add_vectors(current_pos, direction)
                path.append(current_pos)

        return path, obstacles

    def generate_path_info(self, grid_size: tuple, params: dict) -> PathInfo: # type: ignore
        print("    LOG: Generating 'hub_with_stepped_islands' topology...")

        hub_size = params.get('hub_size', 3)
        island_size = params.get('island_size', 3)
        gap_size = params.get('gap_size', 3)
        height_options = params.get('height_options', [])

        if not height_options or not isinstance(height_options, list):
            height_options = [2, -2, 3, -3] # Mặc định 4 đảo

        num_islands = len(height_options)

        center_x = grid_size[0] // 2
        center_z = grid_size[2] // 2

        placement_coords = []
        obstacles = []
        branches = [] # [MỚI] Để lưu các nhánh cho metadata

        # 1. Tạo hòn đảo trung tâm (Hub)
        hub_center: Coord = (center_x, 0, center_z)
        start_pos = hub_center
        hub_coords = self._create_square_platform(hub_center, hub_size)
        placement_coords.extend(hub_coords)

        # 2. Xác định vị trí và độ cao cho 4 đảo vệ tinh
        directions = [FORWARD_Z, FORWARD_X, BACKWARD_Z, BACKWARD_X] # Bắc, Đông, Nam, Tây
        random.shuffle(height_options)
        if num_islands < len(directions):
            directions = random.sample(directions, num_islands)

        # 3. Vòng lặp tạo cầu thang và các đảo
        for i in range(num_islands):
            direction = directions[i]
            height_diff = height_options[i]

            # Điểm bắt đầu của cầu thang, nằm ở rìa của Hub
            stair_start_point = add_vectors(hub_center, scale_vector(direction, hub_size // 2))

            # Tạo cầu thang và lấy đường đi, vật cản từ nó
            stair_path, stair_obstacles = self._create_staircase(stair_start_point, direction, height_diff, gap_size)
            obstacles.extend(stair_obstacles)

            # Điểm cuối của cầu thang cũng là điểm vào của đảo vệ tinh
            island_entry_point = stair_path[-1] if stair_path else stair_start_point

            # Tạo hòn đảo vệ tinh
            island_center = add_vectors(island_entry_point, scale_vector(direction, island_size // 2))
            island_coords = self._create_square_platform(island_center, island_size)

            # Cập nhật các khối cần đặt
            placement_coords.extend(island_coords)
            placement_coords.extend(stair_path) # Bậc thang cũng là nơi có thể đi được

            # [MỚI] Ghi nhận nhánh này vào metadata
            branch_path = [hub_center, stair_start_point] + stair_path + [island_center]
            branches.append(branch_path)

        # [FIX] Chọn target_pos là center của đảo cuối cùng (xa nhất từ hub)
        # thay vì lấy ngẫu nhiên từ set
        path_coords = list(set(placement_coords))
        
        # Target là center của đảo cuối cùng trong branches
        if branches:
            last_branch = branches[-1]
            target_pos = last_branch[-1]  # island_center của branch cuối
        else:
            target_pos = start_pos

        # [NEW] Calculate island centers for semantic positions
        island_centers = []
        for branch in branches:
            if branch:
                island_centers.append(branch[-1])  # Last coord is island_center
        
        # [NEW] Semantic positions for function_reuse and hub_spoke strategies
        semantic_positions = {
            'hub_center': hub_center,
            'optimal_start': 'hub_center',
            'optimal_end': 'island_0' if island_centers else 'hub_center',
            'valid_pairs': [
                {
                    'name': 'hub_to_island_easy',
                    'start': 'hub_center',
                    'end': 'island_0',
                    'path_type': 'single_island',
                    'strategies': ['function_reuse', 'hub_spoke'],
                    'difficulty': 'EASY',
                    'teaching_goal': 'Simple hub to single island traversal'
                },
                {
                    'name': 'hub_to_far_island_medium',
                    'start': 'hub_center',
                    'end': 'island_last',
                    'path_type': 'cross_hub',
                    'strategies': ['function_reuse', 'island_replication'],
                    'difficulty': 'MEDIUM',
                    'teaching_goal': 'Visit multiple islands with PROCEDURE reuse'
                },
                {
                    'name': 'island_to_island_hard',
                    'start': 'island_0',
                    'end': 'island_last',
                    'path_type': 'full_traversal',
                    'strategies': ['function_reuse', 'hub_spoke'],
                    'difficulty': 'HARD',
                    'teaching_goal': 'Visit all islands with identical patterns'
                }
            ]
        }
        
        # Add island positions dynamically
        for i, center in enumerate(island_centers):
            semantic_positions[f'island_{i}'] = center
        if island_centers:
            semantic_positions['island_last'] = island_centers[-1]

        # [MỚI] Tạo metadata để FunctionPlacer có thể hiểu cấu trúc
        metadata: Dict[str, Any] = {
            "topology_type": "hub_with_stepped_islands",
            "branches": branches,
            "hub": hub_coords,
            "center": hub_center,
            "semantic_positions": semantic_positions,  # [NEW]
        }

        return PathInfo(
            start_pos=start_pos,
            target_pos=target_pos,
            path_coords=path_coords,
            placement_coords=path_coords,
            obstacles=obstacles,
            metadata=metadata
        )

File: packages/academic-map-generator/src/generator/topologies/HubWithSteppedIslands.ts
/**
 * Hub With Stepped Islands Topology (PORTED FROM PYTHON)
 * Creates a central hub island connected to satellite islands at different heights.
 * Islands are connected by staircases that go up or down.
 * Ideal for teaching functions with parameters and 3D navigation.
 */

import { BaseTopology } from './BaseTopology';
import { IPathInfo, Coord } from '../types';

// Direction vectors
const FORWARD_Z: [number, number, number] = [0, 0, 1];  // North
const FORWARD_X: [number, number, number] = [1, 0, 0];  // East
const BACKWARD_Z: [number, number, number] = [0, 0, -1]; // South
const BACKWARD_X: [number, number, number] = [-1, 0, 0]; // West

const addVectors = (a: Coord, b: [number, number, number]): Coord => 
  [a[0] + b[0], a[1] + b[1], a[2] + b[2]];

const scaleVector = (v: [number, number, number], s: number): [number, number, number] =>
  [v[0] * s, v[1] * s, v[2] * s];

export class HubWithSteppedIslandsTopology extends BaseTopology {
  *generatePathInfoVariants(
    gridSize: [number, number, number],
    params: Record<string, any>,
    maxVariants: number
  ): Generator<IPathInfo> {
    let count = 0;

    // Height scenarios: positive = up, negative = down
    const scenarios = [
      [3, -3],              // 2 islands: 1 up, 1 down
      [2, -2, 4],           // 3 islands: mixed heights
      [2, 3, 4],            // 3 islands: all up
      [2, -2, 3, -3],       // 4 islands
      [-2, -3, -4]          // 3 islands: all down
    ];

    for (const scenario of scenarios) {
      if (count >= maxVariants) return;
      yield this.generatePathInfo(gridSize, {
        ...params,
        height_options: scenario
      });
      count++;
    }
  }

  /**
   * Create a square platform centered at the given position.
   */
  private _createSquarePlatform(center: Coord, size: number): Coord[] {
    const [cx, cy, cz] = center;
    const half = Math.floor(size / 2);
    const coords: Coord[] = [];
    
    for (let xOffset = -half; xOffset <= half; xOffset++) {
      for (let zOffset = -half; zOffset <= half; zOffset++) {
        coords.push([cx + xOffset, cy, cz + zOffset]);
      }
    }
    return coords;
  }

  /**
   * Create a staircase from start_point in direction with height_diff.
   * Positive height_diff = going up, negative = going down.
   * Each step moves 1 horizontal unit and 1 vertical unit.
   * @param gapSize - Additional flat blocks at the end
   */
  private _createStaircase(
    startPoint: Coord, 
    direction: [number, number, number], 
    heightDiff: number, 
    gapSize: number
  ): { pathCoords: Coord[]; obstacles: any[] } {
    const pathCoords: Coord[] = [];
    const obstacles: any[] = [];
    let currentPos: Coord = [...startPoint];

    // Create staircase with 1-height steps (jumpable)
    const stepsNeeded = Math.abs(heightDiff);
    const yStep = heightDiff > 0 ? 1 : -1;

    for (let i = 0; i < stepsNeeded; i++) {
      // Move horizontally
      currentPos = addVectors(currentPos, direction);
      // Change height
      currentPos = [currentPos[0], currentPos[1] + yStep, currentPos[2]];
      pathCoords.push([...currentPos]);
    }

    // Add gap blocks at final height (no Y change)
    const remainingGap = gapSize - stepsNeeded;
    for (let i = 0; i < remainingGap && remainingGap > 0; i++) {
      currentPos = addVectors(currentPos, direction);
      pathCoords.push([...currentPos]);
    }

    return { pathCoords, obstacles };
  }

  generatePathInfo(gridSize: [number, number, number], params: Record<string, any>): IPathInfo {
    console.log("    LOG: Generating 'hub_with_stepped_islands' topology...");

    const hubSize = params.hub_size || 3;
    const islandSize = params.island_size || 3;
    const gapSize = params.gap_size || 3;
    
    // Height options: each value is the height difference for that island
    let heightOptions: number[] = params.height_options;
    if (!heightOptions || !Array.isArray(heightOptions) || heightOptions.length === 0) {
      heightOptions = [2, -2, 3, -3]; // Default: 4 islands with mixed heights
    }

    const numIslands = heightOptions.length;

    // Center the hub in the grid
    const centerX = params.center_x || Math.floor(gridSize[0] / 2);
    const centerZ = params.center_z || Math.floor(gridSize[2] / 2);

    const placementCoords: Coord[] = [];
    const obstacles: any[] = [];
    const branches: Coord[][] = [];

    // 1. Create central hub
    const hubCenter: Coord = [centerX, 0, centerZ];
    const startPos = hubCenter;
    const hubCoords = this._createSquarePlatform(hubCenter, hubSize);
    placementCoords.push(...hubCoords);

    // 2. Define directions for satellite islands (N, E, S, W)
    const allDirections = [FORWARD_Z, FORWARD_X, BACKWARD_Z, BACKWARD_X];
    
    // Shuffle height options for randomness
    const shuffledHeights = [...heightOptions].sort(() => Math.random() - 0.5);
    
    // Use only as many directions as we have islands
    const directions = numIslands < allDirections.length 
      ? this._shuffleArray([...allDirections]).slice(0, numIslands)
      : allDirections;

    // 3. Create staircases and islands
    const islandCenters: Coord[] = [];
    
    for (let i = 0; i < numIslands; i++) {
      const direction = directions[i % directions.length];
      const heightDiff = shuffledHeights[i];

      // Start point of staircase: edge of hub
      const stairStartPoint = addVectors(hubCenter, scaleVector(direction, Math.floor(hubSize / 2)));

      // Create staircase
      const { pathCoords: stairPath, obstacles: stairObstacles } = 
        this._createStaircase(stairStartPoint, direction, heightDiff, gapSize);
      obstacles.push(...stairObstacles);

      // Entry point of island (end of staircase)
      const islandEntryPoint = stairPath.length > 0 ? stairPath[stairPath.length - 1] : stairStartPoint;

      // Create satellite island
      const islandCenter = addVectors(islandEntryPoint, scaleVector(direction, Math.floor(islandSize / 2)));
      const islandCoords = this._createSquarePlatform(islandCenter, islandSize);
      islandCenters.push(islandCenter);

      // Add to placement coords
      placementCoords.push(...islandCoords);
      placementCoords.push(...stairPath);

      // Record branch for metadata
      const branchPath: Coord[] = [hubCenter, stairStartPoint, ...stairPath, islandCenter];
      branches.push(branchPath);
    }

    // Deduplicate placement_coords (all walkable tiles)
    const dedupedPlacement = this._deduplicateCoords(placementCoords);

    // Target is center of last island
    const targetPos = branches.length > 0 
      ? branches[branches.length - 1][branches[branches.length - 1].length - 1]
      : startPos;

    // Compute path_coords using BFS pathfinding
    // This is the SHORTEST path from start to target through walkable tiles
    const pathCoords = this.computePathCoords(startPos, targetPos, dedupedPlacement);

    // Semantic positions (matching Python)
    const semantic_positions: Record<string, any> = {
      hub_center: hubCenter,
      optimal_start: 'hub_center',
      optimal_end: islandCenters.length > 0 ? 'island_0' : 'hub_center',
      valid_pairs: [
        {
          name: 'hub_to_island_easy',
          start: 'hub_center',
          end: 'island_0',
          path_type: 'single_island',
          strategies: ['function_reuse', 'hub_spoke'],
          difficulty: 'EASY',
          teaching_goal: 'Simple hub to single island traversal'
        },
        {
          name: 'hub_to_far_island_medium',
          start: 'hub_center',
          end: 'island_last',
          path_type: 'cross_hub',
          strategies: ['function_reuse', 'island_replication'],
          difficulty: 'MEDIUM',
          teaching_goal: 'Visit multiple islands with PROCEDURE reuse'
        },
        {
          name: 'island_to_island_hard',
          start: 'island_0',
          end: 'island_last',
          path_type: 'full_traversal',
          strategies: ['function_reuse', 'hub_spoke'],
          difficulty: 'HARD',
          teaching_goal: 'Visit all islands with identical patterns'
        }
      ]
    };

    // Add island positions dynamically
    for (let i = 0; i < islandCenters.length; i++) {
      semantic_positions[`island_${i}`] = islandCenters[i];
    }
    if (islandCenters.length > 0) {
      semantic_positions['island_last'] = islandCenters[islandCenters.length - 1];
    }

    // Segments for analysis
    const allSegments: Coord[][] = [hubCoords, ...branches];
    const lengths = allSegments.map(s => s.length);
    const segment_analysis = {
      num_segments: allSegments.length,
      lengths,
      types: ['hub', ...branches.map((_, i) => `spoke_${i}`)],
      min_length: lengths.length > 0 ? Math.min(...lengths) : 0,
      max_length: lengths.length > 0 ? Math.max(...lengths) : 0,
      avg_length: lengths.length > 0 ? lengths.reduce((a, b) => a + b, 0) / lengths.length : 0
    };

    return {
      start_pos: startPos,
      target_pos: targetPos,
      path_coords: pathCoords,           // DYNAMIC: shortest path
      placement_coords: dedupedPlacement, // STATIC: all walkable tiles
      obstacles: obstacles,
      metadata: {
        topology_type: 'hub_with_stepped_islands',
        branches: branches,
        hub: hubCoords,
        hub_center: hubCenter,
        island_centers: islandCenters,
        height_options: heightOptions,
        num_spokes: numIslands,
        segments: allSegments,
        segment_analysis,
        semantic_positions
      }
    };
  }

}
